### PR TITLE
Generate code using CStringHolder instead of String

### PR DESCRIPTION
**Pull request by philn**
Generated code relying on CStringHolder should be more efficient because it
reduces the amount of string copies.

### CODE REVIEW

**Comment by philn** [comment]
CI fails on Pango for which I have a branch. Along with most of gtk-rs modules for which bindings need to be regenerated, I have branches for most of those at this point. I can open PRs for them now or after this one + the glib one is merged. Please advise ;)

**Comment by EPashkin** [comment]
Seems it use need fixed as 'use glib::cstringingholder;`. Just add it to all imports (for glib it will be removed automatically).
```
error[E0433]: failed to resolve. Use of undeclared type or module `cstringholder`
  --> D:/eap/rust/0/gio\src\auto\settings_schema_source.rs:38:57
   |
38 |     pub fn list_schemas(&self, recursive: bool) -> (Vec<cstringholder::CStringHolder>, Vec<cstringholder::CStringHolder>) {
   |                                                         ^^^^^^^^^^^^^ Use of undeclared type or module `cstringholder`
```
Also adding it unneeded in pango\src\auto\coverage.rs, atk\src\auto\socket.rs, gio\src\auto\settings_backend.rs, gio\src\auto\subprocess_launcher.rs

**Comment by philn** [comment]
Ah, I didn't test gio with the `v2_40` feature indeed, I'll fix it. I wanted to avoid "unused" warnings so I tried to import the module only when it's needed in the generated code.

**Comment by philn** [comment]
gio-rs regen there, https://github.com/philn/gio/commit/ea028eb4d0f819345b3ea62ff8fc5352b4b2ff1d

**Comment by philn** [comment]
PR updated!

File: src/codegen/return_value.rs [hunk 1]
```diff
@@ -8,14 +8,17 @@ use traits::*;
...
 }
 
 impl ToReturnValue for library::Parameter {
-    fn to_return_value(&self, env: &Env) -> String {
+    fn to_return_value(&self, env: &Env, is_trampoline: bool) -> String {
         let rust_type =
             parameter_rust_type(env, self.typ, self.direction, self.nullable, RefMode::None);
-        let name = rust_type.into_string();
+        let mut name = rust_type.into_string();
+        if is_trampoline && self.direction == library::ParameterDirection::Return && name == "gstring::GString" {
```
Comment (EPashkin) [inline src/codegen/return_value.rs]: "GString"

File: src/codegen/return_value.rs [hunk 2]
```diff
@@ -8,14 +8,17 @@ use traits::*;
...
+            name = std::string::String::from("String");
```
Comment (EPashkin) [inline src/codegen/return_value.rs]: "String".to_owned() or .into()

File: src/library.rs [hunk 3]
```diff
@@ -277,6 +277,14 @@ impl TypeId {
...
         TypeId { ns_id: 0, id: 33 }
     }
+
+    pub fn tid_gstrv() -> TypeId {
+        TypeId { ns_id: 0, id: 34 }
+    }
+
+    pub fn tid_utf8_list() -> TypeId {
+        TypeId { ns_id: 0, id: 258 }
+    }
```
Comment (EPashkin) [inline src/library.rs]: How these 2 types looks in .gir file?
Property `ScaleButton::get_property_icons(&self) -> Vec<GString>` don't detected by it :(
So maybe its don't have constant id and need be added in Library::new after fundamentals.

**Comment by EPashkin** [comment]
@philn After this paths, you can remove all manual imports, at minimum on gtk-rs 😉 
https://gist.github.com/EPashkin/4780324a59f986be68028f152ce7cf07

**Comment by philn** [comment]
Awesome! I'll check this tomorrow, thanks!

File: src/library.rs (hunk 3 above)
Comment (philn) [reply src/library.rs]: The icons property has the following type in the gir file:
```
        <array>
          <type name="utf8"/>
        </array>
```

Comment (philn) [reply src/library.rs]: I already see `Type::c_array(&mut library, TypeId::tid_utf8(), None, None);` in Library::new()  :/

Comment (EPashkin) [reply src/library.rs]: Strange that you get this big number then,
if this really same type, then `Library::new` need assert.
PS. And what about second added type?

Comment (philn) [reply src/library.rs]: I don't remember for which module I added `gstrv` support actually. I commented it out and no new error appeard :confused:

**Comment by philn** [comment]
PR updated, much nicer thanks to your patch @EPashkin :)

**Comment by EPashkin** [comment]
@philn Now it really much shorter and without new warning. Thanks.
Can you add this patch as final touch?
It use `&GString::from_glib_borrow` in trampolines
```diff
+++ b/src/analysis/trampoline_parameters.rs
@@ -137,6 +137,7 @@ pub fn analyze(
 
         let conversion_type = {
             match *env.library.type_(par.typ) {
+                library::Type::Fundamental(library::Fundamental::Utf8) |
                 library::Type::Record(..) |
                 library::Type::Interface(..) |
                 library::Type::Class(..) => ConversionType::Borrow,
```

**Comment by philn** [comment]
Done. Thanks for the reviews and help

**Comment by EPashkin** [comment]
@philn Thanks again, Lets merge to allow update glib's PR.
Don't forget use `glib/Gir_GObject.toml` too 😉

**Pull request merged**